### PR TITLE
App 90 : Create AnalyticsEventFactory

### PR DESCRIPTION
**Pull request by Drjacky**
## Create AnalyticsEventFactory

[JIRA Issue: APP-90](https://safeboda.atlassian.net/browse/APP-90)

#### Description
Encapsulate all necessary fields and method in AnalyticsEvent and create a factory

#### Task status

| Status |
| ------ |
| CLOSES |

#### How to test it manually
The usage is like this:
```
class MyCustomAnalyticsEventManager {

    companion object {
        fun welcomeScreen() = AnalyticsEventFactory.createAnalyticsEvent(WELCOME, emptyList())
    }
```

### CODE REVIEW

File: app/src/main/java/com/safeboda/commons/analytics/entity/AnalyticsEvent.kt [hunk 1]
```diff
@@ -7,11 +7,19 @@ abstract class AnalyticsEvent(
     val name: String
 ) {
 
-    abstract fun getProperties(): Map<String, AnalyticsValue<out Any>>
+    private val _ldProperties = mutableMapOf<String, AnalyticsValue<out Any>>()
```
Comment (salvamr) [inline app/src/main/java/com/safeboda/commons/analytics/entity/AnalyticsEvent.kt]: What "lb" stands for?

**Comment by salvamr** [comment]
image

I'm getting this error when trying to use it in a sample project

**Comment by Drjacky** [comment]
> image
> 
> I'm getting this error when trying to use it in a sample project

Maybe I'm wrong but I guess you're using the factory incorrectly! Even before this change by me, the way to pass data and use `setUser` was like creating a custom class that extends `AnalyticsUser`, then using that custom class like:
```
private fun createCustomAnalyticsUser(
        params: SendMetaDataParams,
        metaData: MetaData
    ): CustomAnalyticsUser {
        return SafeBodaAnalyticsUser(
            id = params.user.id,
            identifier = params.user.phoneNumber,
            firstName = params.user.firstName,
            email = params.user.email ?: ""
        )
    }
```
🤔

**Comment by salvamr** [comment]
> > image
> > I'm getting this error when trying to use it in a sample project
> 
> Maybe I'm wrong but I guess you're using the factory incorrectly! Even before this change by me, the way to pass data and use `setUser` was like creating a custom class that extends `AnalyticsUser`, then using that custom class like:
> 
> ```
> private fun createCustomAnalyticsUser(
>         params: SendMetaDataParams,
>         metaData: MetaData
>     ): CustomAnalyticsUser {
>         return SafeBodaAnalyticsUser(
>             id = params.user.id,
>             identifier = params.user.phoneNumber,
>             firstName = params.user.firstName,
>             email = params.user.email ?: ""
>         )
>     }
> ```
> 
> 🤔

Ok, totally my bad. I used the factory in a wrong way. Thanks! 😅

**Comment by salvamr** [comment]
image
All good now. I was using the wrong AnalyticsService method ...

File: app/src/main/java/com/safeboda/commons/analytics/entity/AnalyticsEvent.kt [hunk 2]
```diff
@@ -7,11 +7,19 @@ abstract class AnalyticsEvent(
...
+    }
 
-    fun toJsonObject() = getProperties().mapToJsonObject()
+    protected fun <T : Any> addListOfProperties(name: String, values: List<T>) {
+        _properties[name] = AnalyticsValue(values = values)
+    }
+
+    internal fun toBundle() = _properties.mapToBundle()
+
+    internal fun toJsonObject() = _properties.mapToJsonObject()
```
Comment (mapm14) [inline app/src/main/java/com/safeboda/commons/analytics/entity/AnalyticsEvent.kt]: Please keep this functions `public`

**Pull request merged**